File: android/src/main/kotlin/com/axun/healthsdk/pc300_health_sdk/Pc300HealthSdkPlugin.kt
package com.axun.healthsdk.pc300_health_sdk

import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.Log
import androidx.annotation.NonNull
import androidx.annotation.RequiresApi
import com.axun.healthsdk.pc300_health_sdk.model.BlueDevice
import com.axun.healthsdk.pc300_health_sdk.utils.BlueManageUtils
import com.axun.healthsdk.pc300_health_sdk.utils.toAccessorJson
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.PluginRegistry.Registrar

/** Pc300HealthSdkPlugin */
public class Pc300HealthSdkPlugin : FlutterPlugin, MethodCallHandler ,ActivityAware{
    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        this.flutterPluginBinding = flutterPluginBinding
        BlueManageUtils.instance.init(flutterPluginBinding.applicationContext)
//        BluetoothContext.set(flutterPluginBinding.applicationContext)

    }

    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        flutterPluginBinding = null
    }

    override fun onDetachedFromActivity() {
        BlueManageUtils.instance.release()
        channel?.setMethodCallHandler(null)
        channel = null

    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {

    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        instance.channel = MethodChannel(flutterPluginBinding?.binaryMessenger, "pc300_health_sdk")
       instance.channel ?.setMethodCallHandler(Pc300HealthSdkPlugin())
    }

    override fun onDetachedFromActivityForConfigChanges() {

    }
    // This static function is optional and equivalent to onAttachedToEngine. It supports the old
    // pre-Flutter-1.12 Android projects. You are encouraged to continue supporting
    // plugin registration via this function while apps migrate to use the new Android APIs
    // post-flutter-1.12 via https://flutter.dev/go/android-project-migration.
    //
    // It is encouraged to share logic between onAttachedToEngine and registerWith to keep
    // them functionally equivalent. Only one of onAttachedToEngine or registerWith will be called
    // depending on the user's project. onAttachedToEngine or registerWith must both be defined
    // in the same class.
    companion object {

        const val TAG = "Pc300HealthSdkPlugin"
        val instance by lazy(LazyThreadSafetyMode.SYNCHRONIZED) {
            Pc300HealthSdkPlugin()
        }


        @JvmStatic
        fun registerWith(registrar: Registrar) {
            BlueManageUtils.instance.init(registrar.context())
            instance.channel  = MethodChannel(registrar.messenger(), "pc300_health_sdk")
            instance.channel ?.setMethodCallHandler(Pc300HealthSdkPlugin())
        }


        val blueDeviceListener = object : BlueManageUtils.OnBlueToothCallback {
            override fun onStartDiscovery() {
                Log.d("blueSearchStatus===》","开始搜索")
            }

            @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
            override fun OnDiscoveryCompleted(devices: MutableList<BluetoothDevice>) {

                val list: MutableList<BlueDevice> = ArrayList()
                devices.forEach {
                    val item = BlueDevice(address = it.address,
                            name = it.name, type = it.type, bondState = it.bondState)
                    list.add(item)
                }
                sendChannelMessage(onDiscoveryCompleteCode,list.toAccessorJson())

            }

            override fun onConnectSuccess() {
                val map = HashMap<String, Any>()
                map["success"] = true
                map["message"] = "success"
                sendChannelMessage(onConnectSuccessCode, map)
            }

            override fun onConnectError(error: String) {
                val map = HashMap<String, Any>()
                map["success"] = false
                map["message"] = error
                sendChannelMessage(onConnectErrorCode, map)
            }


            @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
            override fun onFindDevice(device: BluetoothDevice) {
                Log.d("66666",device.name)
                val item = BlueDevice(address = device.address,
                        name = device.name, type = device.type, bondState = device.bondState)
                sendChannelMessage(onFindDeviceCode,item.toAccessorJson())
            }

        }


        fun sendChannelMessage(messageCode: Int, value: Any?) {
            val message = Message()
            message.what = messageCode
            message.obj = value;
            handler.sendMessage(message)
        }

        var handler = @SuppressLint("HandlerLeak")
        object : Handler(Looper.getMainLooper()) {
            override fun handleMessage(msg: Message?) {
                super.handleMessage(msg)
                when (msg?.what) {
                    onConnectSuccessCode -> {
                        instance.channel?.invokeMethod("onConnectSuccess", msg.obj)
                    }

                    onConnectErrorCode -> {
                        instance.channel?.invokeMethod("onConnectError", msg.obj)
                    }

                    /*获取设备id*/
                    onGetDeviceIDCode -> {
                        instance.channel?.invokeMethod("onGetDeviceID",msg.obj)
                    }
                    /*获取到设备版本信息*/
                    onGetDeviceVerCode -> {
                        instance.channel?.invokeMethod("onGetDeviceVer",msg.obj)
                    }
                    /*获取到心电模块版本*/
                    onGetECGVerCode -> {
                        instance.channel?.invokeMethod("onGetECGVer",msg.obj)
                    }
                    /*获取到血氧参数*/
                    onGetSpO2ParamCode -> {
                        instance.channel?.invokeMethod("onGetSpO2Param",msg.obj)
                    }
                    /*获取到血氧波形数据*/
                    onGetSpO2WaveCode -> {
                        instance.channel?.invokeMethod("onGetSpO2Wave",msg.obj)
                    }
                    /*血压测量状态改变*/
                    onGetNIBPActionCode -> {
                        instance.channel?.invokeMethod("onGetNIBPAction",msg.obj)
                    }

                    /*获取到实时袖带压力值*/
                    onGetNIBPRealTimeCode -> {
                        instance.channel?.invokeMethod("onGetNIBPRealTime",msg.obj)
                    }

                    /*血压测量结果*/
                    onGetNIBPResultCode -> {
                        instance.channel?.invokeMethod("onGetNIBPResult",msg.obj)
                    }
                    /*心电测量状态改变*/
                    onGetECGActionCode -> {
                        instance.channel?.invokeMethod("onGetECGAction",msg.obj)
                    }
                    /*获取心电实时数据*/
                    onGetECGRealTimeCode -> {
                        instance.channel?.invokeMethod("onGetECGRealTime",msg.obj)
                    }
                    /*心电测量结果*/
                    onGetECGResultCode -> {
                        instance.channel?.invokeMethod("onGetECGResult",msg.obj)
                    }
                    /*获取到体温数据*/
                    onGetTmpCode -> {
                        instance.channel?.invokeMethod("onGetTmp",msg.obj)
                    }
                    /*获取到血糖值*/
                    onGetGluCode -> {
                        instance.channel?.invokeMethod("onGetGlu",msg.obj)
                    }
                    /*获取到血压模块状态*/
                    onGetNIBPStatusCode -> {
                        instance.channel?.invokeMethod("onGetNIBPStatus",msg.obj)
                    }
                    /*获取到血氧模块状态*/
                    onGetSpO2StatusCode -> {
                        instance.channel?.invokeMethod("onGetSpO2Status",msg.obj)
                    }
                    /*获取到血糖模块状态*/
                    onGetGluStatusCode -> {
                        instance.channel?.invokeMethod("onGetGluStatus",msg.obj)
                    }
                    /*获取到体温模块状态*/
                    onGetTmpStatusCode -> {
                        instance.channel?.invokeMethod("onGetTmpStatus",msg.obj)
                    }
                    /*下位机关机*/
                    onGetPowerOffCode -> {
                        instance.channel?.invokeMethod("onGetPowerOff",msg.obj)
                    }
                    /*与设备连接丢失*/
                    onConnectLoseCode -> {
                        instance.channel?.invokeMethod("onConnectLose",msg.obj)
                    }
                    /*获取全部设备*/
                    onDiscoveryCompleteCode -> {
                        instance.channel?.invokeMethod("onDiscoveryComplete",msg.obj)
                    }

                    /*查询到设备*/
                    onFindDeviceCode->{

                        instance.channel?.invokeMethod("onFindDevice",msg.obj)
                    }

                }
            }
        }

        /*连接成功*/
        val onConnectSuccessCode = 0x100
        /*连接失败*/
        val onConnectErrorCode = 0x101

        /*获取全部设备列表*/
        val onDiscoveryCompleteCode = 0x121

        val onFindDeviceCode = 0x122

        /*获取到设备ID*/
        val onGetDeviceIDCode = 0x102

        /*获取到设备版本信息*/
        val onGetDeviceVerCode = 0x103

        /*获取到心电模块版本*/
        val onGetECGVerCode = 0x104

        /*获取到血氧参数*/
        val onGetSpO2ParamCode = 0x105

        /*获取到血氧波形数据*/
        val onGetSpO2WaveCode = 0x106

        /*血压测量状态改变*/
        val onGetNIBPActionCode = 0x107

        /*血压测量结果*/
        val onGetNIBPResultCode = 0x108

        /*心电测量状态改变*/
        val onGetECGActionCode = 0x109

        /*获取心电实时数据*/
        val onGetECGRealTimeCode = 0x110

        /*心电测量结果*/
        val onGetECGResultCode = 0x111

        /*获取到体温数据*/
        val onGetTmpCode = 0x112

        /*获取到血糖值*/
        val onGetGluCode = 0x113

        /*获取到血压模块状态*/
        val onGetNIBPStatusCode = 0x114

        /*获取到血氧模块状态*/
        val onGetSpO2StatusCode = 0x115

        /*获取到血糖模块状态*/
        val onGetGluStatusCode = 0x116

        /*获取到体温模块状态*/
        val onGetTmpStatusCode = 0x117

        /*下位机关机*/
        val onGetPowerOffCode = 0x118

        /*与设备连接丢失*/
        val onConnectLoseCode = 0x119

        /*获取到实时袖带压力值*/
        val onGetNIBPRealTimeCode = 0x120
    }


    private var channel: MethodChannel? = null
    private var flutterPluginBinding:FlutterPlugin.FlutterPluginBinding?=null



    @RequiresApi(Build.VERSION_CODES.JELLY_BEAN_MR2)
    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "getPlatformVersion" -> {
                result.success("Android ${android.os.Build.VERSION.RELEASE}")
            }

            /**
             * @Description 检查蓝牙是否打开
             **/
            "isOpen" -> {
                val isOpen = BlueManageUtils.instance.client.isOpen
                result.success(isOpen)
            }

            /**
             * @Description 打开蓝牙
             **/
            "openDevice" -> {
                val open = BlueManageUtils.instance.client.Open()
                result.success(open)
            }

            /**
             * @Description 关闭蓝牙
             **/
            "closeDevice" -> {
                val close = BlueManageUtils.instance.client.Close()
                result.success(close.toString())
            }

            /**
             * @Description 获取已绑定设备
             **/
            "getBondedDevices" -> {
                val devices = BlueManageUtils.instance.client.bondedDevices
                val list: MutableList<BlueDevice> = ArrayList()

                devices.forEach {
                    val item = BlueDevice(address = it.address,
                            name = it.name, type = it.type, bondState = it.bondState)
                    list.add(item)
                }
                result.success(list.toAccessorJson())
            }

            /**
             * @Description 连接设备
             **/
            "connect" -> {
                val address = call.argument<String>("address")
                val device = BluetoothAdapter.getDefaultAdapter().getRemoteDevice(address)
                BlueManageUtils.instance. client.StopDiscovery()
                BlueManageUtils.instance.client.Connect(device)
            }

            /**
             * @Description 断开连接
             **/
            "disConnect" -> {
                BlueManageUtils.instance.release()
            }

            /**
             * @Description 开始搜索设备
             **/
            "startDiscovery"->{
                BlueManageUtils.instance.startDiscovery(blueDeviceListener)
            }

            /**
             * @Description 开始接受数据
             **/
            "startMeasure" -> {
                BlueManageUtils.instance.healthClient.Start()
            }

            /**
             * @Description 停止接受数据
             **/
            "stopMeasure" -> {
                BlueManageUtils.instance.healthClient.Stop()
            }

            /**
             * @Description 暂停接受数据
             **/
            "pauseMeasure" -> {
                BlueManageUtils.instance.healthClient.Pause()
            }

            /**
             * @Description 恢复接受数据
             **/
            "continueMeasure" -> {
                BlueManageUtils.instance.healthClient.Continue()
            }

            /**
             * @Description 查询设备版本信息
             **/
            "queryDeviceVer" -> {
                BlueManageUtils.instance.healthClient.QueryDeviceVer()
            }

            /**
             * @Description 查询血压模块状态
             **/
            "queryNIBPStatus" -> {
                BlueManageUtils.instance.healthClient.QueryNIBPStatus()
            }

            /**
             * @Description 查询血氧模块状态
             **/
            "querySpO2Status" -> {
                BlueManageUtils.instance.healthClient.QuerySpO2Status()
            }

            /**
             * @Description 查询血糖模块状态
             **/
            "queryGluStatus" ->{
                BlueManageUtils.instance.healthClient.QueryGluStatus()
            }

            /**
             * @Description 查询体温模块状态
             **/
            "queryTmpStatus"->{
                BlueManageUtils.instance.healthClient.QueryTmpStatus()
            }

            /**
             * @Description 查询心电模块版本信息
             **/
            "queryECGVer" -> {
                BlueManageUtils.instance.healthClient.QueryECGVer()
            }

            /**
             * @Description 血压测量控制
             **/
            "setNIBPAction" -> {
                val startMeasure = call.argument<Boolean>("startMeasure")
                BlueManageUtils.instance.healthClient.SetNIBPAction(startMeasure?:false)
            }

            /**
             * @Description 心电测量控制
             **/
            "setECGMotion"->{
                val startMeasure = call.argument<Boolean>("startMeasure")
                BlueManageUtils.instance.healthClient.SetECGMotion(startMeasure?:false)
            }

            else -> {
                result.notImplemented()
            }
        }
    }






}

File: ios/Classes/Pc300HealthSdkPlugin.m
#import "Pc300HealthSdkPlugin.h"
#import "BlueManageUtils.h"

@implementation Pc300HealthSdkPlugin
{
    FlutterMethodChannel *methodChannel;
    BlueManageUtils *manageUtils;
}
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
    FlutterMethodChannel* channel = [FlutterMethodChannel
                                     methodChannelWithName:@"pc300_health_sdk"
                                     binaryMessenger:[registrar messenger]];
    Pc300HealthSdkPlugin* instance = [[Pc300HealthSdkPlugin alloc] initWithChannel:channel registrar:registrar messenger:[registrar messenger]];
    [registrar addMethodCallDelegate:instance channel:channel];
}

- (instancetype)initWithChannel:(FlutterMethodChannel *)channel
                      registrar:(NSObject<FlutterPluginRegistrar>*)registrar
                      messenger:(NSObject<FlutterBinaryMessenger>*)messenger {
    
    if (self) {
        self->methodChannel = channel;
        self->manageUtils =  [BlueManageUtils shareEngine];
    }
    
    return self;
}

- (void)handleMethodCallWithName:(NSString *)methodName withParm:(NSDictionary *)map
{
    if ([methodName isEqualToString:onConnectSuccess])
    {
        ///链接成功
        NSDictionary *parm = @{
            @"success": @YES,
            @"message": @"success"
        };
        [self->methodChannel invokeMethod:@"onConnectSuccess" arguments:[self dataToJsonString:parm]];
    }
    else if([methodName isEqualToString:onConnectError])
    {
        NSDictionary *parm = @{
            @"success": @NO,
            @"message": @"error"
        };
        [self->methodChannel invokeMethod:@"onConnectError" arguments:[self dataToJsonString:parm]];
    }
    else if([methodName isEqualToString:OnGetNIBPResult])
    {
        [self->methodChannel invokeMethod:@"onGetNIBPResult" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetECGAction])
    {
        [self->methodChannel invokeMethod:@"onGetECGAction" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetNIBPRealTime])
    {
        [self->methodChannel invokeMethod:@"onGetNIBPRealTime" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetECGRealTime])
    {
        [self->methodChannel invokeMethod:@"onGetECGRealTime" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetNIBPStatus])
    {
        [self->methodChannel invokeMethod:@"onGetNIBPStatus" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetPowerOff])
    {
        [self->methodChannel invokeMethod:@"onGetPowerOff" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetECGResult])
    {
        [self->methodChannel invokeMethod:@"onGetECGResult" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetDeviceID])
    {
        [self->methodChannel invokeMethod:@"onGetDeviceID" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetGlu])
    {
        [self->methodChannel invokeMethod:@"onGetGlu" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetNIBPAction])
    {
        [self->methodChannel invokeMethod:@"onGetNIBPAction" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetSpO2Status])
    {
        [self->methodChannel invokeMethod:@"onGetSpO2Status" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetDeviceVer])
    {
        [self->methodChannel invokeMethod:@"onGetDeviceVer" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetTmpStatus])
    {
        [self->methodChannel invokeMethod:@"onGetTmpStatus" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetTmp])
    {
        [self->methodChannel invokeMethod:@"onGetTmp" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetECGVer])
    {
        [self->methodChannel invokeMethod:@"onGetECGVer" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetSpO2Wave])
    {
        [self->methodChannel invokeMethod:@"onGetSpO2Wave" arguments:[self dataToJsonString:map]];
    }
    else if([methodName isEqualToString:OnGetSpO2Param])
    {
        [self->methodChannel invokeMethod:@"onGetSpO2Param" arguments:[self dataToJsonString:map]];
    }
}
- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result
{
    if ([@"getPlatformVersion" isEqualToString:call.method])
    {
        result([@"iOS " stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
    }
    else if ([@"isOpen" isEqualToString:call.method])
    {
        /**
         *
         *  * 蓝牙状态
         */
        result(@([self->manageUtils isOpen]));
    }
    else if ([@"startDiscovery" isEqualToString:call.method])
    {
        /**
         *
         *  * 搜索蓝牙设备
         */
        float time = 10;
        if (call.arguments[@"maxTime"] != nil) time = [call.arguments[@"maxTime"] floatValue];
        
        [self->manageUtils startScan:time scanResult:^(NSMutableArray * _Nonnull deviceList) {
            NSLog(@"_methodChannel - 打印%@",self->methodChannel);
            NSLog(@"_manageUtils - 打印%@",self->manageUtils);
            [self->methodChannel invokeMethod:@"onDiscoveryComplete" arguments:[self dataToJsonString:deviceList]];
        }];
    }
    else if ([@"connect" isEqualToString:call.method])
    {
        /**
         *
         *  * 连接设备
         */
        NSString *address = call.arguments[@"address"];
        [self->manageUtils connectDevice:address];
    }
    else if ([@"disConnect" isEqualToString:call.method])
    {
        /**
         *
         *  * 断开连接
         */
        [self->manageUtils disconnectDevice];
    }
    else if ([@"getBondedDevices" isEqualToString:call.method])
    {
        /**
         *
         *  * 获取设备列表
         */
        result([self dataToJsonString:[self->manageUtils GetDeviceList]]);
    }
    else if ([@"setNIBPAction" isEqualToString:call.method])
    {
        /**
         * @Description 血压测量控制
         **/
        bool startMeasure = call.arguments[@"startMeasure"];
        [self->manageUtils SetNIBPAction:startMeasure];
    }
    else if ([@"setECGMotion" isEqualToString:call.method])
    {
        /**
         * @Description 心电测量控制
         **/
        bool startMeasure = call.arguments[@"startMeasure"];
        [self->manageUtils SetECGAction:startMeasure];
    }
    else if ([@"stopMeasure" isEqualToString:call.method])
    {
        /**
         * @Description 停止接受数据
         **/
        [[CRSpotCheck sharedInstance] closeTimerSpotCheck];
    }
    else if ([@"continueMeasure" isEqualToString:call.method])
    {
        /**
         * @Description 恢复接受数据
         **/
        [[CRSpotCheck sharedInstance] restartTimerSpotCheck];
    }
    else if ([@"queryDeviceVer" isEqualToString:call.method])
    {
        ///查询设备版本信息
        [self->manageUtils QueryDeviceVer];
    }
    else if ([@"queryNIBPStatus" isEqualToString:call.method])
    {
        ///查询血压模块状态
        [self->manageUtils QueryNIBPStatus];
    }
    else if ([@"querySpO2Status" isEqualToString:call.method])
    {
        ///查询血氧模块状态
        [self->manageUtils QuerySpO2Status];
    }
    else if ([@"queryGluStatus" isEqualToString:call.method])
    {
        ///查询血糖模块状态
        [self->manageUtils QueryGluStatus];
    }
    else if ([@"queryTmpStatus" isEqualToString:call.method])
    {
        ///查询体温模块状态
        [self->manageUtils QueryTmpStatus];
    }
    else if ([@"queryECGVer" isEqualToString:call.method])
    {
        ///查询心电模块版本信息
        [self->manageUtils QueryEcgVer];
    }
    else
    {
        result(FlutterMethodNotImplemented);
    }
}

///data转json字符串
- (NSString *)dataToJsonString:(id)obj
{
    NSError *parseError = nil;
    NSData *jsonData = [NSJSONSerialization dataWithJSONObject:obj options:NSJSONWritingPrettyPrinted error:&parseError];
    NSString *jsonString = [[NSString alloc] initWithData:jsonData encoding:NSUTF8StringEncoding];
    return jsonString;
}
@end
